### PR TITLE
Move validation from Controllers into Service Layer

### PR DESCRIPTION
**Pull request by brownbearsoftware**
Just wanting to get some early feedback on the pattern before proceeding.

Still need to do some unit tests for Departures/remove some integration tests + move the other types into service layer too.

### CODE REVIEW

**Review by pezholio — DISMISSED** [review]
Yeah, this is a really nice approach. Happy with this 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/ValidatableActionResult.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.model
+
+sealed interface ValidatableActionResult<ResultType> {
+  data class Success<ResultType>(val result: ResultType) : ValidatableActionResult<ResultType>
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/ValidatableActionResult.kt:4]: I think `entity` and `EntityType` would be better names here, i.e:

```suggestion
  data class Success<EntityType>(val entity: EntityType) : ValidatableActionResult<EntityType>
```

**Review by pezholio — DISMISSED** [review]
Looks great, thanks so much for this 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt [hunk 2]
```diff
@@ -291,7 +291,7 @@ class PremisesController(
   }
 
   private fun <ResultType> extractResultEntityOrThrow(result: ValidatableActionResult<ResultType>) = when (result) {
-    is ValidatableActionResult.Success -> result.result
+    is ValidatableActionResult.Success -> result.entity
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:294]: Could you squash these commits into the others? I'm a big fan of https://github.com/tummychow/git-absorb for this. No worries if it causes a big mess.

**Review by pezholio — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt (hunk 2 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt]: sure thing, have sorted

**Review by pezholio — APPROVED** [review]

**Pull request merged**